src/NServiceBus.Core/Transports/Learning/AsyncFile.cs

```diff
@@ -95,5 +95,61 @@ static FileStream CreateReadStream(string filePath)
         {
             return new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read, bufferSize: 4096, useAsync: true);
         }
+
+        public static async Task<bool> Move(string source, string target)
+        {
+            try
+            {
+                File.Move(source, target);
+            }
+            catch (IOException)
+            {
+                return false;
+            }
+
+            var count = 0;
+            while (IsFileLocked(new FileInfo(target)))
+            {
+                await Task.Delay(100).ConfigureAwait(false);
+
+                count++;
+
+                if (count > 10)
+                {
+                    break;
+                }
+            }
+
+            //seem like File.Move is not atomic at least within the same process so we need this extra check
+            return File.Exists(target);
```

Comment:
according to ms:

> If you are performing multiple operations on the same file, it can be more efficient to use FileInfo instance methods instead of the corresponding static methods of the File class, because a security check will not always be necessary.

if we're already creating a file info, we should use that for all the file operations I think?